pkg/labels/labels_test.go

```diff
@@ -629,3 +630,30 @@ func TestBuilder_Labels(t *testing.T) {
 		}).Labels(),
 	)
 }
+
+func TestLabels_Hash(t *testing.T) {
+	lbls := Labels{
+		{Name: "foo", Value: "bar"},
+		{Name: "baz", Value: "qux"},
+	}
+	testutil.Equals(t, lbls.Hash(), lbls.Hash())
+	testutil.Assert(t, lbls.Hash() != Labels{lbls[1], lbls[0]}.Hash(), "unordered labels match.")
+	testutil.Assert(t, lbls.Hash() != Labels{lbls[0]}.Hash(), "different labels match.")
+}
+
+var benchmarkLabels_Result uint64
+
+func BenchmarkLabels_Hash(b *testing.B) {
+	b.ReportAllocs()
+
+	lbl := &strings.Builder{}
+	lbl.Grow(1024 * 1024 * 10) // 10MB.
```

Comment:
This is an unrealistic test, how does it look with a more typical values?

Reply:
lol for label bytes in total being less than 1024 bytes (the allocated b), `Hash` seems to be fully on stack 0.o

```
/tmp/___BenchmarkLabels_Hash_in_github_com_prometheus_prometheus_pkg_labels -test.v -test.bench ^\QBenchmarkLabels_Hash\E$ -test.run ^$
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/pkg/labels
BenchmarkLabels_Hash
BenchmarkLabels_Hash-12    	 6806316	       169 ns/op	       0 B/op	       0 allocs/op
PASS
``` 
If I have at least 1 byte more, it's worse than my solution: (10x alloc, 4x latency) 0.o

Reply:
For 
```
func BenchmarkLabels_Hash(b *testing.B) {
	b.ReportAllocs()

	lbls := make(Labels, 9)
	// Typical series of 10 labels.
	for i := 0; i < len(lbls); i++ {
		// Typical label ~50B name, 50B value.
		lbls[i] = Label{Name: fmt.Sprintf("abcdefghijabcdefghijabcdefghijabcdefghijabcdefghij%d", i), Value: fmt.Sprintf("abcdefghijabcdefghijabcdefghijabcdefghijabcdefghij%d", i)}
	}
	var h uint64

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		h = lbls.Hash2()
	}
	benchmarkLabelsResult = h
}
```
It's  `BenchmarkLabels_Hash-12    	 6242728	       186 ns/op	       0 B/op	       0 allocs/op`

If we change labels to 10 it's `BenchmarkLabels_Hash-12    	 2396838	       904 ns/op	    2048 B/op	       1 allocs/op` ;p in the same time my code: 
`BenchmarkLabels_Hash-12    	 1577956	       682 ns/op	     100 B/op	      21 allocs/op`

Reply:
This is some crazy ass optimization compiler does (: And increasing size of `b` to 1MB is obviously automatically moving this to heap.

Reply:
The vast majority of labelpairs are going to be under 1KB, only info metrics might vary there. So the current code is probably best then.

Reply:
This is really sketchy, how sure we are that this optimization stays like this in real deployment? It's extremely fragile. It goes to heap and suddenly we allocate constantly extreme amount of Bs

Reply:
We call this fairly rarely in the tsdb (only when a new series comes along), so I'm not too worried. We allocate a pile of other things when that happens too, so the net effect would be small.

Reply:
> We call this fairly rarely in the tsdb (only when a new series comes along),

hmmmm, really? I can see it used every request (e.g for Thanos Receive maybe)

Reply:
Yes, it's only on the new series codepath. The whole Prometheus 2.x scraping design avoids calling this frequently.

Reply:
Yes we don't have an equivalent fast path like Prometheus does. I hacked up a PoC once, but never ended up implementing it the whole way (it was essentially doing the same as Prometheus where we cut out the bytes describing the labels in the remote write proto and had a similar cache to Prometheus for fast series lookups).

That said, why wouldn't we want to make this path as fast as possible for Prometheus? I don't see a downside to the "slow path" being as fast as possible.

Reply:
As I understand it the new fast path would be slower than the current fast path, which doesn't seem like a good tradeoff given that the slow path should be very rare.

Reply:
Ah I didn't understand it that way. I agree with your statement though, the fast path should not get slower.

Reply:
Yup, I will play with it to ensure we don't sacrifice fast path

Reply:
I made it work efficiently for both fast and slow path. PTAL